cmds/games/READMEs/ski.man



SKI(6)                   GAMES AND DEMOS                   SKI(6)



NAME
     ski - an unusual skiing simulation game

SYNOPSIS
     ski

DESCRIPTION
     This document describes Ski!  Version  6.0  of  October  14,
     1990.

     Imagine you are skiing down an infinite slope,  facing  such
     hazards  as  trees,  ice,  bare ground, and the Snoman! -- a
     horrible monster!   Each  turn  moves  you  one  screen  row
     downhill.   Unfortunately,  you have to put your jet-powered
     skis on backwards, so you can't  see  ahead  where  you  are
     going; only behind where you have been.

     However, you can turn to either side, jump  or  hop  through
     the  air, teleport through hyperspace, launch nuclear ICBMs,
     and cast spells to call  the  Fire  Demon.   And  since  the
     hazards  occur  in  patches, you can skillfully out-maneuver
     them.

     You ski until something bad  happens  to  you.   Your  final
     score  is  based  on how far you have skiied, how many jumps
     you made, how many Snomen have melted during your  run,  and
     how healthy you are when are finished.

     These characters represent the things that you  see  on  the
     slope behind you:

     I    This is where you are.

     .    This is normal snow -- always completely  safe  to  ski
          on.

     Y    These are trees.  If you try to ski through trees,  you
          might hit one and get hurt!

     ' '  A space represents bare ground with no snow on it.   If
          you try to ski on dirt, you could fall down....

     #    This is ice.  When you are on ice, you can't turn,  and
          you might slip and fall down.

     A    This is the Snoman!  As soon as  he  crawls  out  of  a
          Snobank,  he  will  lumber  directly towards you at one
          space per turn, and if he gets within one space of you,
          he  will  grab  you  and  beat you up!  However, he has
          trouble moving through trees.   Steer  clear  from  the
          Snoman,  and  he will melt after a while, since it is a
          sunny day.  If you jump over him, your jet-powered skis



Sun Release 4.1   Last change: 14 October 1990                  1






SKI(6)                   GAMES AND DEMOS                   SKI(6)



          will help melt him.  You can also melt him with nuclear
          ICBMs, or call the Fire Demon who will  cook  him,  but
          once  a Snoman melts, another one always shows up after
          a while.  The Snoman doesn't understand about the edges
          of  the slope either, so you can just cross them to get
          away from him!  And since he is slower  than  you,  you
          can try to ski right past him before he can grab you!

     *    This represents a nuclear ICBM that you launched.   You
          launch  them from your backpack, and they zoom right at
          the Snoman.

     D    This is the Fire  Demon,  who  is  always  your  friend
          unless you call him a bad name and he gets mad at you.

     These are your commands.  You can  type  any  one  of  these
     letters when you have the "?" prompt on the end of the slope
     line, or just hit a return to do nothing this turn.

     R    Turn to the right.  If you were  going  straight  down,
          this  will  make  you  go  one space to the right every
          turn.  If you type "R" again,  then  you  will  go  two
          spaces  to  the  right every turn.  If you type "L" the
          next turn, you will again be going  one  space  to  the
          right  every  turn.   Remember that you can't turn when
          you are on ice, and if you go off the edge of the slope
          you  come back into the other edge.  You travel farther
          while you are going sideways, so you get a better score
          if  you  slalom to the side a lot, but you can never go
          sideways more than five spaces per turn.

     L    Turn left.  Exactly the  same  as  turning  right,  but
          totally opposite.

     J    Jump.  When jumping, you don't  have  the  "?"  on  the
          right for each line you jump over, so you can't enter a
          command for it.  You can't  fall,  hit  trees,  or  get
          grabbed by the Snoman while jumping, but you might land
          badly and get hurt.  Since the Snoman moves  one  space
          toward you each turn, you have to be going at least two
          spaces per turn sideways to jump over him without  get-
          ting  grabbed  when you land, but your jet-powered skis
          may melt the Snoman while you're jumping over him.

     H    Hop.  A hop is half a jump.

     T    Teleport through hyperspace.  This  transports  you  to
          somewhere  else on the slope, but you might materialize
          way up in the air and fall down.

     I    Launch ICBM with nuclear warhead.  If your backpack  is
          working  well, an ICBM will be launched from it towards



Sun Release 4.1   Last change: 14 October 1990                  2






SKI(6)                   GAMES AND DEMOS                   SKI(6)



          the  Snoman,  travelling  at  three  spaces  per  turn.
          Launching ICBMs is dangerous, but once one is going, it
          can't hurt you.   If  it  is  launched  and  loses  its
          negative-infrared   lock   on   the  Snoman,  it  self-
          destructs.

     D    Cast spell to bring the Fire  Demon  to  this  material
          plane.  If the alternate universes are in a synchronous
          phase, the Fire Demon  will  appear  somewhere  on  the
          slope,  and  move to attack the Snoman at one space per
          turn, but if something goes wrong  while  casting  this
          spell,  the  Fire  Demon might get mad and decide to go
          for you instead of the Snoman!  If there is  no  Snoman
          around, the Fire Demon will soon get bored and go home.
          If you don't like where the Fire Demon showed  up,  you
          can  try  to  recast this spell, so that he will appear
          somewhere else.

     Correspondence: resumix!stevans@decwrl.dec.com

     Good luck, and happy skiing!

AUTHOR
     Mark Stevans, for Roy's Games, Inc.

DIAGNOSTICS
     None.

BUGS
     None known.

























Sun Release 4.1   Last change: 14 October 1990                  3





cmds/games/ski.c
/*                                  S K I !
**
**                       Version 6.0 of October 14, 1990
**
**           Copyright 1983-1990 by Mark Stevans for Roy's Games, Inc.
*/

#include <stdio.h>
#include <ctype.h>
#include "ski.h"

/*
**
** Constants controlling the multiple cellular growth automatons that are
** executed in parallel to generate hazards.
**
** Each cellular growth automaton probability element tends to control a
** different facet of hazard generation:
**
**    [0] appearance of new hazards in clear snow
**    [1] hazards edge growth
**    [2] hazards edge stability
**    [3] appearance of holes in solid hazards (this allows the Snoman to
**        work his way through forests)
*/

PRIVATE float prob_tree[] = { 0.0, 30.0, 70.0, 90.0 };
PRIVATE float prob_ice[] = { 0.0, 30.0, 70.0, 90.0 };
PRIVATE float prob_ground[] = { 0.0, 30.0, 70.0, 90.0 };
PRIVATE float prob_snoman_appearance;

/*
** "init" performs various initialization functions.
*/

PRIVATE int init(level_ptr)
LEVEL *level_ptr;
{
	register int i;

	/*
	** Print the game version.
	*/

	printf("SKI!  Version 6.0!\n");

	/*
	** Initialize the random number generator.
	*/

	RANDOM_INITIALIZE((int) time(NIL(long *)));

	/*
	** Initialize the player's position.
	*/

	level_ptr->player_pos = RANDOM_GENERATE() % LINE_LEN;

	/*
	** Initialize the current level.
	*/

	for (i = 0; i < LINE_LEN; ++i)
		level_ptr->slope[i] = REP_SNOW;

	level_ptr->slope[LINE_LEN] = EOS;

	level_ptr->meters_travelled = 0;
	level_ptr->jump_count = -1;
	level_ptr->level_num = 0;
	level_ptr->num_snomen_melted = 0;
	level_ptr->num_jumps_attempted = 0.0;
	level_ptr->player_pos = RANDOM_GENERATE() % LINE_LEN;
	level_ptr->snoman_pos = NO_POSITION;
	level_ptr->icbm_pos = NO_POSITION;
	level_ptr->demon_pos = NO_POSITION;
	level_ptr->player_speed = 0;

	/*
	** Randomize the appearance probabilities.
	*/

	prob_tree[0] = (RANDOM_GENERATE() % 100) / 500.0 + 0.05;
	prob_ice[0] = (RANDOM_GENERATE() % 100) / 500.0 + 0.05;
	prob_ground[0] = (RANDOM_GENERATE() % 100) / 500.0 + 0.05;
	prob_snoman_appearance = (RANDOM_GENERATE() % 100) / 25.0 + 1.0;

	/*
	** Return a code indicating successful completion.
	*/

	return (SUCCESS);
}

/*
** "main" is the top level function of the game.
*/

PUBLIC int main()
{
	LEVEL level;

	/*
	** Initialize the game.
	*/

	if (init(&level) != SUCCESS)
		exit(1);

	/*
	** Perform the game loop until the game is over.
	*/

	for (;;) {
		draw_picture(&level);
		do_user(&level);
		manipulate_objects(&level);
		update_level(&level);
	}
}

/*
** "gen_next_slope" generates the slope of the next level, dependent upon
** the characteristics of the current level, the probabilities, and the
** position of the player.
*/

PRIVATE int gen_next_slope(current_level, next_level)
LEVEL *current_level;
LEVEL *next_level;
{
	register int i;
	register char *current_slope;
	register char *next_slope;
	register int player_pos;
	short num_nearby_trees, num_nearby_ice, num_nearby_ground;

	/*
	** Cache some convenient values.
	*/

	current_slope = current_level->slope;
	next_slope = next_level->slope;
	player_pos = current_level->player_pos;

	/*
	** Generate each character of the next level.
	*/

	for (i = 0; i < LINE_LEN; ++i) {
		/*
		** Count the number of nearby trees, ice patches, and
		** ground patches on the current level.
		*/

		num_nearby_trees = 0;
		num_nearby_ice = 0;
		num_nearby_ground = 0;

		if (current_slope[i] == REP_TREE)
			++num_nearby_trees;
		if (current_slope[i] == REP_ICE)
			++num_nearby_ice;
		if (current_slope[i] == REP_GROUND)
			++num_nearby_ground;

		if (i > 0) {
			if (current_slope[i - 1] == REP_TREE)
				++num_nearby_trees;
			else if (current_slope[i - 1] == REP_ICE)
				++num_nearby_ice;
			else if (current_slope[i - 1] == REP_GROUND)
				++num_nearby_ground;
		}

		if (i < (LINE_LEN - 1)) {
			if (current_slope[i + 1] == REP_TREE)
				++num_nearby_trees;
			else if (current_slope[i + 1] == REP_ICE)
				++num_nearby_ice;
			else if (current_slope[i + 1] == REP_GROUND)
				++num_nearby_ground;
		}

		/*
		** Generate this character of the next level based upon
		** the characteristics of the nearby characters on the
		** current level.
		*/

		if (PROB(prob_tree[num_nearby_trees]) &&
				((i != player_pos) || (num_nearby_trees > 0)))
			next_slope[i] = REP_TREE;
		else if (PROB(prob_ice[num_nearby_ice]) &&
				((i != player_pos) || (num_nearby_ice > 0)))
			next_slope[i] = REP_ICE;
		else if (PROB(prob_ground[num_nearby_ground]) &&
				((i != player_pos) || (num_nearby_ground > 0)))
			next_slope[i] = REP_GROUND;
		else
			next_slope[i] = REP_SNOW;
	}

	/*
	** Terminate the slope string.
	*/

	next_slope[LINE_LEN] = EOS;

	/*
	** Return a code indicating successful completion.
	*/

	return (SUCCESS);
}

/*
** "update_level" moves to the next level.
*/

PRIVATE int update_level(current_level)
LEVEL *current_level;
{
	LEVEL next_level;

	/*
	** Go to the next level, and move the player.
	*/

	++current_level->level_num;

	current_level->meters_travelled += ABS(current_level->player_speed) + 1;

	/*
	** Figure out the new player position based on a modulo
	** addition.  Note that we must add the line length into the
	** expression before taking the modulus to make sure that we
	** are not taking the modulus of a negative integer.
	*/

	current_level->player_pos = (current_level->player_pos +
			current_level->player_speed + LINE_LEN) % LINE_LEN;

	/*
	** Generate the updated slope.
	*/

	gen_next_slope(current_level, &next_level);

	strcpy(current_level->slope, next_level.slope);

	/*
	** If the player was jumping, decrement the jump count.
	*/

	if (current_level->jump_count >= 0)
		--current_level->jump_count;

	/*
	** If there is no Snoman, one might be created.
	*/

	if (!EXISTS(current_level->snoman_pos)) {
		if (PROB(prob_snoman_appearance)) {
			/*
			** Make sure that the Snoman does not appear too
			** close to the player.
			*/

			do {
				current_level->snoman_pos = RANDOM_GENERATE() % LINE_LEN;
			} while (ABS(current_level->snoman_pos -
					current_level->player_pos) <=
					MIN_SNOMAN_APPEARANCE_DISTANCE);
		}
	}

	/*
	** Increase the initial appearance probabilities of all obstacles and
	** the Snoman.
	*/

	prob_tree[0] *= LEVEL_MULTIPLIER;
	prob_ice[0] *= LEVEL_MULTIPLIER;
	prob_ground[0] *= LEVEL_MULTIPLIER;
	prob_snoman_appearance *= LEVEL_MULTIPLIER;

	/*
	** Return a code indicating successful completion.
	*/

	return (SUCCESS);
}
